### PR TITLE
add 10.0-mrp_product_use_case

### PR DESCRIPTION
**Pull request by tonygalmiche**
add 10.0-mrp_product_use_case

### CODE REVIEW

**Comment by pedrobaeza** [comment]
You can't use submodules here.

**Comment by tonygalmiche** [comment]
how to do ?

**Comment by pedrobaeza** [comment]
Well, you have to copy full module on the branch.

**Comment by tonygalmiche** [comment]
I do not master the branch system with github
So suddenly, I do not understand what I have to do
This is my first module (my second attempt)

**Comment by pedrobaeza** [comment]
A proposal:

- Fork the repository
- Start a new branch
- Copy your module in that branch
- Commit the changes
- Push here

**Comment by tonygalmiche** [comment]
That's what I did :
git clone git@github.com:tonygalmiche/manufacture.git
cd manufacture/
git checkout 10.0
git checkout -b 10.0-mrp_product_use_case
copy / past my addon
git add .
git commit -a
git push --set-upstream origin 10.0-mrp_product_use_case
Button  « New pull request » on github

where is my mistake?

**Comment by pedrobaeza** [comment]
You didn't copy your module, but it seems only a relative link to it.

**Comment by tonygalmiche** [comment]
I understood
I did not delete the .git file from my addon

**Comment by pedrobaeza** [comment]
Now the module is there, but it doesn't comply the needed guidelines. You have to check https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md and adapt your code conveniently, but maybe it's better to put first what is the goal of the module to see if it's interesting for other people, it overlaps any other feature here in OCA, etc.

**Comment by tonygalmiche** [comment]
The problem is that I am so bad in English (not easy to contribute on OCA).
David BEAL of Akretion who wanted me to offer this module because he needs it for a project.
Then if I manage to contribute to OCA I think I will have other modules on the production to propose.

**Comment by tonygalmiche** [comment]
I do not understand this problem because I changed the code as it should be done : 
=> E8103(sql-injection)

**Comment by tonygalmiche** [comment]
Where is 5 errors ?
=> pylint expected errors 0, found 5!

**Comment by tonygalmiche** [comment]
Where is my error ?

**Comment by tonygalmiche** [comment]
Why test_flake8 => FAIL ?

**Comment by tonygalmiche** [comment]
it's all ok :)

**Comment by tonygalmiche** [comment]
what to do for that : 
- Some checks were not successful - 2 failing and 2 successful checks
-> codecov/patch — 46.66% of diff hit (target 92.32%)
-> codecov/project — 89.72% (-2.61%) compared to ea0c547 

I do not see what I have to do.
I do not understand.

File: mrp_product_use_case/README.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+* Tony Galmiche <tony.galmiche@infosaone.com> (http://infosaone.com)
+
+Do not contact contributors directly about support or help with technical issues.
+
+Funders
+-------
+
+The development of this module has been financially supported by:
+
+* Open Source Integrators <http://www.opensourceintegrators.com>
```
Comment (bealdav) [inline mrp_product_use_case/README.rst]: Sure ?

**Review by bealdav — APPROVED** [review]
Not sure License file is required, license key in __manifest__.py is ok. 
Please save the planet, remove useless bytes ;-)

Maybe you can squash some commits.

**Comment by pedrobaeza** [comment]
I still don't know what this module does.

**Comment by tonygalmiche** [comment]
I add usage in README.rst and image 

but this syntax does not work : 
MRP Product Use Case

how to add an image ?

**Comment by bealdav** [comment]
https://github.com/OCA/mis-builder/blob/10.0/mis_builder/README.rst

https://raw.githubusercontent.com/OCA/mis-builder/10.0/mis_builder/README.rst

File: mrp_product_use_case/security/ir.model.access.csv [hunk 2]
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+
```
Comment (bealdav) [inline mrp_product_use_case/security/ir.model.access.csv]: please remove this file

File: mrp_product_use_case/wizard/mrp_product_use_case_wizard.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            self.env['mrp.product.use.case.line'].create(vals)
+            self.mrp_product_use_case(wizard_id, compose_id, level + 1)
+
+    @api.multi
+    def do_search_component(self):
+        for obj in self:
+            if obj.product_id:
+                self.mrp_product_use_case(obj.id, obj.product_id.id, 1)
+        return {
+            'name': u"MRP Product Use Case " + str(obj.product_id.name),
```
Comment (bealdav) [inline mrp_product_use_case/wizard/mrp_product_use_case_wizard.py]: small improvement replace 
"MRP Product Use Case" by "Use Case" for a better display

**Comment by tonygalmiche** [comment]
I can not put a full url if the module is not published?

`.. figure:: https://raw.githubusercontent.com/OCA/mis-builder/10.0/mis_builder/static/description/ex_report_settings.png`

**Comment by bealdav** [comment]
it's not a problem, it will displayed when published

**Comment by bealdav** [comment]
======== Testing test_flake8 ========
mrp_product_use_case/__manifest__.py:9:80: E501 line too long (110 > 79 characters)

File: mrp_product_use_case/README.rst [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+====================
+
+This module add MRP Product Use Case functionality
+
+
+Usage
+=====
+
+* From the form of a component click on the button 'Use Case'.
+* or click on the 'Master Data / Product Use Case' and choise a component.
```
Comment (bealdav) [inline mrp_product_use_case/README.rst]: s/choise/choose

File: mrp_product_use_case/README.rst [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+* The result displays the list of finished products containing this component and in which level of the bom this component appears
```
Comment (bealdav) [inline mrp_product_use_case/README.rst]: cut the line

File: mrp_mto_with_stock/README.rst [hunk 6]
```diff
@@ -29,9 +29,10 @@ To configure this module, you need to:
    standard behavior.
 
 If you want to use the second mode, based on forecast quantity
+
 #. Go to the warehouse you want to follow this behaviour.
 #. In the view form go to the tab *Warehouse Configuration* and set the 
-   *MRP MTO with forecast stock*. You still need to configure the products
+   MRP MTO with forecast stock. You still need to configure the products
```
Comment (LoisRForgeFlow) [inline mrp_mto_with_stock/README.rst]: This is a fix in another module, this most of the times should be a different PR, but since it's just the README let me just ask you to re-add the `*`'s, they are used to format the text between them as italics. The blank line you added at L32 is good!

File: mrp_product_use_case/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Tony Galmiche - InfoSaône <tony.galmiche@infosaone.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'MRP Product Use Case',
+    'version': '10.0.1.0.0',
+    'category': 'mrp',
+    'summary': 'This module allows to know the case of multi-level use of a component from the nomenclatures',
```
Comment (LoisRForgeFlow) [inline mrp_product_use_case/__manifest__.py]: This lines is exceeding the 80 characters limit, please break it to have a happy travis :smile:

**Comment by bealdav** [comment]
ping @lreficent

File: mrp_product_use_case/wizard/mrp_product_use_case_wizard.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            self.env['mrp.product.use.case.line'].create(vals)
+            self.mrp_product_use_case(wizard_id, compose_id, level + 1)
+
+    @api.multi
+    def do_search_component(self):
+        for obj in self:
+            if obj.product_id:
+                self.mrp_product_use_case(obj.id, obj.product_id.id, 1)
+        return {
+            'name': u"Use Case " + str(obj.product_id.name),
```
Comment (bealdav) [inline mrp_product_use_case/wizard/mrp_product_use_case_wizard.py]: Prefer 
`'name': "Use Case %s" % obj.product_id.name,`
to 
`'name': u"Use Case " + str(obj.product_id.name),`

File: mrp_product_use_case/wizard/mrp_product_use_case_wizard.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class MrpProductUseCaseLine(models.TransientModel):
+    _name = 'mrp.product.use.case.line'
+
+    wizard_id = fields.Many2one('mrp.product.use.case.wizard', 'Wizard')
+    level = fields.Char('Level')
+    component_id = fields.Many2one('product.product', 'Component')
+    line = fields.Integer('Line')
+    quantity = fields.Float('Quantity', digits=(14, 6))
```
Comment (hparfr) [inline mrp_product_use_case/wizard/mrp_product_use_case_wizard.py]: Why specifiing digits instead of using get_precision ?

Comment (tonygalmiche) [reply mrp_product_use_case/wizard/mrp_product_use_case_wizard.py]: it's corrected

File: mrp_product_use_case/wizard/mrp_product_use_case_wizard.py (hunk 8 above)
Comment (tonygalmiche) [reply mrp_product_use_case/wizard/mrp_product_use_case_wizard.py]: it's corrected

**Review by hparfr — APPROVED** [review]

**Comment by hparfr** [comment]
Code review. Didn't test LGTM.

File: mrp_product_use_case/README.rst [hunk 10]
```diff
@@ -0,0 +1,72 @@
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+   :target: https://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+====================
+MRP Product Use Case
```
Comment (LoisRForgeFlow) [inline mrp_product_use_case/README.rst]: IMHO the name is not very clear, something like `mrp_bom_product_find` would be more descriptive of the functionality, I find "Use Case" very misleading.

File: mrp_product_use_case/README.rst [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+====================
+
+This module add MRP Product Use Case functionality
```
Comment (LoisRForgeFlow) [inline mrp_product_use_case/README.rst]: You should explain here what is "the product use case functionality". For instance "Allows you to find the product usages across all Bill of Materials".

File: mrp_product_use_case/README.rst [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ 
+
+.. image:: static/description/mrp_product_use_case.png
+    :alt: Sample Product Use Case
+    :width: 95 %
+    :align: center
+
+
+Known issues / Roadmap
+======================
```
Comment (LoisRForgeFlow) [inline mrp_product_use_case/README.rst]: Remove empty setion

File: mrp_product_use_case/README.rst [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+* or click on the 'Master Data / Product Use Case' and choose a component.
+* The result displays the list of finished products containing this
+  component and in which level of the bom this component appears
+ 
+
+.. image:: static/description/mrp_product_use_case.png
+    :alt: Sample Product Use Case
+    :width: 95 %
+    :align: center
+
```
Comment (LoisRForgeFlow) [inline mrp_product_use_case/README.rst]: Add the link to runbot on the usage section:

```
.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
   :alt: Try me on Runbot
   :target: https://runbot.odoo-community.org/runbot/{repo_id}/{branch}

.. repo_id is available in https://github.com/OCA/maintainer-tools/blob/master/tools/repos_with_ids.txt
.. branch is "8.0" for exampl
```
(see [oca template](https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst))

File: mrp_product_use_case/__manifest__.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Tony Galmiche - InfoSaône <tony.galmiche@infosaone.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'MRP Product Use Case',
+    'version': '10.0.1.0.0',
+    'category': 'mrp',
+    'summary': """This module allows to know the case of multi-level use
+of a component from the nomenclatures""",
```
Comment (LoisRForgeFlow) [inline mrp_product_use_case/__manifest__.py]: break this line in 2, it is making travis to be red.

File: mrp_product_use_case/README.rst (hunk 10 above)
Comment (bealdav) [reply mrp_product_use_case/README.rst]: Yes name can be improved as suggested or

mrp_component_use_case or mrp_bom_component_use_case or mrp_component_find

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]
Thanks a lot for you contribution. Let me make some suggestions/questions.

image

Why are you showing all the structure? and in that case, the leveling you are displaying is a little bit strange IMHO, why those dashes (-2, --3)?. 

The top product of a Product structure should be the level 0 (In the picture above it would be the [MANUF] Top) the next one should be level 1 (subproduct 1), level 2 (subproduct 1-1) and so on...

Sorry if I look nitpicking and thanks again for contributing to OCA.

File: mrp_product_use_case/README.rst (hunk 10 above)
Comment (LoisRForgeFlow) [reply mrp_product_use_case/README.rst]: I like mrp_component_find, because this is exactly what the module is: a wizard to find components on the manufacturing app (mrp).

Comment (LoisRForgeFlow) [reply mrp_product_use_case/README.rst]: I see that there is a module called `mrp_bom_component_menu` already in OCA so I would follow that structure as well.  What do you thing of `mrp_bom_component_find` for this module?

File: mrp_product_use_case/__manifest__.py (hunk 14 above)
Comment (bealdav) [reply mrp_product_use_case/__manifest__.py]: s / nomenclatures / bom

File: mrp_product_use_case/README.rst (hunk 10 above)
Comment (bealdav) [reply mrp_product_use_case/README.rst]: +1 for mrp_bom_component_find

Comment (bealdav) [reply mrp_product_use_case/README.rst]: @pedrobaeza what do you think of this name ?

**Comment by tonygalmiche** [comment]
@lreficent 
thanks for your corrections

The dashes in the level column allow to have a graphical representation of the level:
```

1
1
-2
-2
--3
-2
...
```

**Comment by tonygalmiche** [comment]
it is done :
  mrp_product_use_case => mrp_bom_component_find


Le 20/04/2018 à 08:48, David Beal a écrit :

File: mrp_bom_component_find/README.rst [hunk 15]
```diff
@@ -0,0 +1,74 @@
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+   :target: https://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+====================
+MRP Product Use Case
```
Comment (bealdav) [inline mrp_bom_component_find/README.rst]: Please you have to rename all strings Use Case if useless in your module without forget one (in screenshot too). 
Sorry and thanks

**Comment by tonygalmiche** [comment]
It's done


> Please you have to rename all strings Use Case if useless in your 
> module without forget one (in screenshot too).
>
> Sorry and thanks
>

**Comment by bealdav** [comment]
Nice for these improvement :+1:

**Comment by bealdav** [comment]
ping @lreficent

**Review by LoisRForgeFlow — APPROVED** [review]
Great, thanks for the fixes :+1:

**Comment by tonygalmiche** [comment]
Thank you @lreficent for the merge
Thank you all for your help @bealdav, @pedrobaeza, @lreficent and @hparfr
